docs/rfcs/027-pageserver-migration.md

```diff
@@ -0,0 +1,454 @@
+is also still attached, and in use by endpoints, then this layer deletion could
+lead to a loss of availability for the endpoint when reading from the previous
+generation pageserver.
+
+The _AttachedMulti_ state simply disables deletions. These will be enqueued
+until the control plane eventually transitions the node into AttachedSingle,
+which unblocks deletions.
+
+AttachedMulti is not required for data safety, only to preserve availability
+on pageservers running with stale generations.
+
+A node enters AttachedMulti only when explicitly asked to by the control plane.
+
+#### AttachedStale
+
+Currently, a pageserver with a stale generation number will continue to
+upload layers, but be prevented from completing deletions. This is safe, but inefficient: layers uploaded by this stale generation
+will not be read back by future generations of pageservers.
+
+The _AttachedStale_ state simply disables S3 uploads. The stale pageserver
+will continue to ingest the WAL and write layers to local disk, but not to
+do any uploads to S3.
+
+A node may enter AttachedStale in two ways:
+
+- Explicitly, when control plane calls into the node at the start of a migration.
+- Implicitly, when the node tries to validate some deletions and discovers
+  that its generation is tale.
+
+Over long periods of time, a node in AttachedStale will accumulate data
+on local disk, as it cannot evict any layers written since it entered the
+AttachStale state. We rely on the control plane to revert the location to
+Secondary or Detached at the end of a migration.
+
+### Warm secondary updates
+
+#### Layer heatmap
+
+The secondary location's job is to serve reads **with the same quality of service as the original location
+was serving them around the time of a migration**. This does not mean the secondary
+location needs the whole set of layers: inactive layers that might soon
+be evicted on the attached pageserver need not be downloaded by the
+secondary. A totally idle tenant only needs to maintain enough on-disk
+state to enable a fast cold start (i.e. the most recent image layers are
+typically sufficient).
+
+To enable this, we introduce the concept of a _layer heatmap_, which
+acts as an advisory input to secondary locations to decide which
+layers to download from S3.
+
+#### Attached pageserver
+
+The attached pageserver, if in state AttachedSingle, periodically
+uploads a serialized heat map to S3. It may skip this if there
+is no change since the last time it uploaded (e.g. if the tenant
+is totally idle).
+
+#### Secondary pageserver behavior
+
+Secondary warm locations run a simple loop, implemented separately from
+the main `Tenant` type, which represents attached tenants:
+
+- Download the layer heatmap
+- Select any "hot enough" layers to download, if there is sufficient
+  free disk space.
+- Download layers
+- Download the latest index_part.json
+- Check if any layers currently on disk are no longer referenced by
+  the tenant's metadata & delete them
```

Comment:
The comparison with index_part.json can be skipped. Any layer not showing up in the heatmap is obviously not worth keeping on the standby. And with my RFC for crash consistent layer map https://github.com/neondatabase/neon/pull/5086 , the attach procedure when we promote secondary to primary will eliminate all non-referenced layers anyways.

Reply:
> Any layer not showing up in the heatmap is obviously not worth keeping on the standby.

This depends on disk pressure.  If a pageserver has plenty of spare disk space, then it might choose to keep layers for longer even if the heatmap doesn't mention them.  For layers that are still in index_part, we should prefer to keep them even if they are cold, until there is some disk pressure to prompt us to remove them.

When the primary maintains the heatmap, it will probably apply some threshold to decide when to drop a layer from the heatmap entirely: that should be decoupled from the secondary pageserver's decision about whether to drop the layer from disk, because the primary doesn't know how much spare disk the secondary has.`

Reply:
Added a paragraph to explain that the heatmap is advisory and the secondary is free to retain layers that aren't in it, hence the check of IndexPart.